config/cfg_lang.ml

```diff
@@ -123,43 +51,62 @@ module Parser = struct
-    | [ ATOM var; EQ; STRING s; RPARENS ] -> (Pred { var; value = String s }, [])
-    | [ ATOM var; EQ; NUMBER n; RPARENS ] -> (Pred { var; value = Number n }, [])
-    | ATOM var :: rest -> (Pred { var; value = String "true" }, rest)
-    | _ ->
+  module Exp = struct
+    let id = function
+      | { pexp_desc = Pexp_ident { txt = Lident "any"; _ }; _ } -> `any
+      | { pexp_desc = Pexp_ident { txt = Lident "all"; _ }; _ } -> `all
+      | { pexp_desc = Pexp_ident { txt = Lident "not"; _ }; _ } -> `not
+      | { pexp_desc = Pexp_ident { txt = Lident "="; _ }; _ } -> `eq
+      | _ -> `invalid
+
+    let arg_list loc = function
+      | [ (Nolabel, { pexp_desc = Pexp_tuple args; _ }) ] -> args
+      | [ (Nolabel, arg) ] -> [ arg ]
+      | _ -> failwith ~loc "Forms any/all/not must parenthesize its arguments"
+
+    let one_arg loc = function
+      | [ (Nolabel, { pexp_desc = Pexp_tuple _; _ }) ] ->
+          failwith ~loc "Not expressions must have a single parameter"
+      | [ (Nolabel, arg) ] -> arg
+      | _ -> failwith ~loc "Forms any/all/not must parenthesize its arguments"
+
+    let var loc = function
+      | Nolabel, { pexp_desc = Pexp_ident { txt = Lident var; _ }; _ } -> var
+      | _ -> failwith ~loc (Format.sprintf "Expected identifier")
+
+    let value loc = function
+      | Nolabel, { pexp_desc = Pexp_constant (Pconst_integer (i, _)); _ } ->
+          Number (int_of_string i)
+      | Nolabel, { pexp_desc = Pexp_constant (Pconst_string (s, _, _)); _ } ->
+          String s
+      | ( Nolabel,
+          { pexp_desc = Pexp_construct ({ txt = Lident "true"; _ }, None); _ } )
+        ->
+          String "true"
+      | _ -> failwith ~loc (Format.sprintf "Expected int, string or bool value")
+
+    let rec parse ~loc (exp : expression) =
+      let fail () =
         failwith ~loc
-          (Format.asprintf "Invalid sequence of tokens: %a" Lexer.pp tokens)
-
-  and parse_list ~loc ?(acc = []) tokens =
-    match tokens with
-    | [] -> (List.rev acc, [])
-    | COMMA :: RPARENS :: rest | RPARENS :: rest -> (List.rev acc, rest)
-    | COMMA :: rest | rest ->
-        let pred, rest = do_parse ~loc rest in
-        parse_list ~loc rest ~acc:(pred :: acc)
+          (Format.asprintf "Invalid expression: %a" Pprintast.expression exp)
+      in
+      match exp.pexp_desc with
+      | Pexp_apply (fn, args) -> (
+          match id fn with
+          | `any -> Any (List.map (parse ~loc) (arg_list loc args))
+          | `all -> All (List.map (parse ~loc) (arg_list loc args))
+          | `not -> Not (parse ~loc (one_arg loc args))
+          | `eq -> (
+              match args with
+              | [ x; y ] -> Pred { var = var loc x; value = value loc y }
+              | _ -> fail ())
+          | `invalid -> fail ())
+      | Pexp_ident { txt = Lident var; _ } ->
+          Pred { var; value = String "true" }
+      | _ -> fail ()
+  end
+
+  let parse = Exp.parse
 end
 
 module Eval = struct
```

Comment:
Below on line 131-133:
https://github.com/ocaml-sys/config.ml/pull/11/files#diff-5ac6a8cf5936a2f2fc4063127c6d907fd023624dddd2072e5b009e00209f2881R131-R133
```ocaml
  let eval ~loc ~env str =
    let parsetree = Parser.parse ~loc str in
    do_eval ~loc ~env parsetree
```

Should we rename the variable `str` to `exp`?

Reply:
Good catch, done.